pvlib/irradiance.py

```diff
@@ -1481,24 +1481,21 @@ def _disc_kn(clearness_index, airmass, max_airmass=12):
 
     am = np.minimum(am, max_airmass)  # GH 450
 
-    # powers of kt will be used repeatedly, so compute only once
-    kt2 = kt * kt  # about the same as kt ** 2
-    kt3 = kt2 * kt  # 5-10x faster than kt ** 3
-
+    # Use Horner's method to compure polynomials efficiently
     bools = (kt <= 0.6)
     a = np.where(bools,
-                 0.512 - 1.56*kt + 2.286*kt2 - 2.222*kt3,
-                 -5.743 + 21.77*kt - 27.49*kt2 + 11.56*kt3)
+        0.512 + kt*(-1.56 + kt*(2.286 - 2.222*kt)),
+        -5.743 + kt*(21.77 + kt*(-27.49 + 11.56*kt)))
```

Comment:
If you want speed, this should be twice as fast:
```suggestion
        a[bools] = 0.512 + kt[bools]*(-1.56 + kt[bools]*(2.286 - 2.222*kt[bools]))
        a[~bools] = -5.743 + kt[~bools]*(21.77 + kt[~bools]*(-27.49 + 11.56*kt[~bools])))
```
Or you could use ```np.piecewise```.

Reply:
I don't actually see a huge change using boolean indices:

```python
>>> %timeit test_disc_value
# 21.2 ns ± 0.619 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

>>> %timeit test_disc_overirradiance
# 19.2 ns ± 0.332 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Reply:
Pity, I would have thought that halving the number of multiplications would make a bigger difference.

Reply:
We should test performance with at least 1000x larger arrays to simulate typical pvlib use cases. Tests that take nanoseconds are often python limited rather than numpy limited.

Reply:
100k is a nice number: ~one year of 5 minute data.

Reply:
FWIW, the current `irradiance.py` benchmarks use 10 days of 1 minute data, so 14400 points. I found this is a good compromise between accuracy and speed for most pvlib functions - it adds up across the whole test suite.

https://github.com/pvlib/pvlib-python/blob/bbc863dc1d0889160d255c2d74bf24e2b98cf08d/benchmarks/benchmarks/irradiance.py#L12-L13

`asv` can be a bit cumbersome, but the instructions for running the benchmarks are not too bad: https://github.com/pvlib/pvlib-python/tree/main/benchmarks

Reply:
ASV results for `time_disc()` show a decrease from about 11ms to 9ms, or about 20%, same as before:
![image](https://user-images.githubusercontent.com/1385621/225435196-8f805a8e-c7f4-4fc7-bba6-7eb5547d8605.png)

TBH, I think the main change is in the airmass polynomial, because the powers were not pre-calculated like `kt2` and `kt3` were for `a`, `b`, and `c`.